### ISSUE TITLE
(JRuby) RuntimeError: You called sig twice without declaring a method in between

### ISSUE THREAD

**Issue by sgbarath619** [Shopify/ruby-lsp#1263]
### Operating System

MacOS

### Ruby version

jruby-9.4.5(ruby 3.1.4)

### Project has a bundle

- [ ] Has bundle

### Ruby version manager being used

rvm

### Description

```
RuntimeError: You called sig twice without declaring a method in between
  _declare_sig_internal at /Users/barathwajs/.rvm/gems/jruby-9.4.5.0/gems/sorbet-runtime-0.5.11142/lib/types/private/methods/_methods.rb:50
            declare_sig at /Users/barathwajs/.rvm/gems/jruby-9.4.5.0/gems/sorbet-runtime-0.5.11142/lib/types/private/methods/_methods.rb:35
                    sig at /Users/barathwajs/.rvm/gems/jruby-9.4.5.0/gems/sorbet-runtime-0.5.11142/lib/types/sig.rb:28
        singleton class at /Users/barathwajs/.rvm/gems/jruby-9.4.5.0/gems/ruby-lsp-0.12.5/lib/core_ext/uri.rb:9
        <class:Generic> at /Users/barathwajs/.rvm/gems/jruby-9.4.5.0/gems/ruby-lsp-0.12.5/lib/core_ext/uri.rb:6
           <module:URI> at /Users/barathwajs/.rvm/gems/jruby-9.4.5.0/gems/ruby-lsp-0.12.5/lib/core_ext/uri.rb:5
                 <main> at /Users/barathwajs/.rvm/gems/jruby-9.4.5.0/gems/ruby-lsp-0.12.5/lib/core_ext/uri.rb:4
                require at org/jruby/RubyKernel.java:1071
                 <main> at /Users/barathwajs/.rvm/gems/jruby-9.4.5.0/gems/ruby-lsp-0.12.5/lib/ruby_lsp/internal.rb:20
                require at org/jruby/RubyKernel.java:1071
                 <main> at /Users/barathwajs/.rvm/gems/jruby-9.4.5.0/gems/ruby-lsp-0.12.5/exe/ruby-lsp:89
                   load at org/jruby/RubyKernel.java:1107
                 <main> at /Users/barathwajs/.rvm/gems/jruby-9.4.5.0/bin/ruby-lsp:25
                   eval at org/jruby/RubyKernel.java:1161
                 <main> at /Users/barathwajs/.rvm/gems/jruby-9.4.5.0/bin/jruby_executable_hooks:22
```

This is the error I am facing when I installed the ruby-lsp extension in vs-code
I am strill a beginner with ruby and ROR, please help me with this

**Comment by andyw8** [comment]
I looked into the current state of JRuby with `sorbet-runtime`.

After cloning the `sorbet`repo, removing `pry` and `pry-byebug` from the `gemspec`, I ran the following against `jruby 9.4.5.0`:

```
cd gems/sorbet-runtime
bundle
bundle rake test
```

which results in:

```
882 runs, 2581 assertions, 15 failures, 48 errors, 9 skips
```

**Comment by nirvdrum** [comment]
@headius Are you aware of any issues in JRuby that might be causing this?

**Comment by headius** [comment]
Without a reproduction it is difficult to know what is happening here, but if I'm reading that file correctly it is seeing a second "sig" being declared without an intervening method definition happening. One possibility would be that the `method_added` hook used to trigger `_on_method_added` is either not firing, or it is firing but does not match expectations needed to consume the current declaration.

I don't know of any specific issues with us calling `method_added`. A reproduction would be helpful here.

**Comment by nirvdrum** [comment]
@headius I think the reproduction is in the Sorbet test suite. But, I don't know the issue well enough. I just thought I'd alert you to it.

**Comment by drkstr101** [comment]
I would like to help sort this out, if possible. We depend on JRuby extensively to integrate with core libraries written for the JVM. Solrgraph has the same issues, FYI.

First, I will see if I can [get Sorbet working smoothly with JRuby](https://github.com/sorbet/sorbet/blob/master/docs/JRuby.md), then loop back here to see what the implications are for ruby-lsp. I have some LSP experience but very limited Ruby so I may need a bit of help there.

**Comment by headius** [comment]
Finally circling back to this today. My assumptions currently:

* When you call "sig" it's setting up a signature for the next method definition.
* The next method definition calls `method_added` which consumes that sig.
* If the `method_added` does not fire or does not consume the sig, a subsequent sig call will error.

Most likely scenario in my mind is that for some reason in this circumstance we're not calling `method_added` or we're calling it in a context that makes it seem like it's not the method the sig was intended for.

**Comment by headius** [comment]
Reproduced locally... shouldn't be long until I have a fix.

```
$ ruby-lsp  
Ruby LSP> Running bundle install for the custom bundle. This may take a while...
Ruby LSP> Command: (bundle check || bundle install) 1>&2
The Gemfile's dependencies are satisfied
bundler: failed to load command: ruby-lsp (/tmp/jruby/3.1.0/bin/ruby-lsp)
RuntimeError: You called sig twice without declaring a method in between
  _declare_sig_internal at /private/tmp/jruby/3.1.0/gems/sorbet-runtime-0.5.11212/lib/types/private/methods/_methods.rb:50
            declare_sig at /private/tmp/jruby/3.1.0/gems/sorbet-runtime-0.5.11212/lib/types/private/methods/_methods.rb:35
                    sig at /private/tmp/jruby/3.1.0/gems/sorbet-runtime-0.5.11212/lib/types/sig.rb:28
```

**Comment by headius** [comment]
Found it...

```
module RubyIndexer
  @configuration = T.let(Configuration.new, Configuration)

  class << self
    extend T::Sig

    sig { returns(Configuration) }
    attr_reader :configuration
  end
end
```

We must not be calling `method_added` for attrs.

Sorbet folks on the thread: it would have been quicker to diagnose this if the `sig` error printed out where the last sig was, since that's the point at which it failed to be claimed.

**Comment by headius** [comment]
Ok it was a bit more subtle than that: we were not calling `singleton_method_added` for attrs on singleton classes. We were still calling `method_added` in that case, which would not trigger the appropriate hook to consume the sig.

Fix in PR now: https://github.com/jruby/jruby/pull/8070

**Comment by headius** [comment]
Thanks @nirvdrum for bringing this to our attention!

For others here: if something doesn't work right the first time in JRuby, assume it's our fault and let us know! It probably is!

**Comment by headius** [comment]
Will be fixed in JRuby 9.4.6.0, which has not been scheduled. I *think* a workaround would be to also define `method_added` alongside `singleton_method_added`, but that should not be changed in Sorbet (it might lead to unexpected results).

Tonight's snapshot should be picked up by tomorrow's setup-ruby job so jruby-head will be fixed by Monday.

**Comment by vinistock** [comment]
Thank you for the assistance @headius! Appreciate the collaboration.

And that's a really good point about the signature error, I'll bring it up with the team to see if we can add the caller location of the sigs.

I'll close this issue since it's already addressed and will soon be released.
